### PR TITLE
Augment the matches

### LINKED ISSUE #4: DOM augmentation
Add badges to proper DOM nodes.

### PR DESCRIPTION
**Pull request by ngotchac**
Closes #4 

See it in action : https://youtu.be/aJmo7O8gzOg

Add an icon on the matches that resolves to an Ethereum Address. Hovering shows the certified badges, and clicking opens an _Account Card_ which shows the address, the badges and the tokens of the corresponding account.

### CODE REVIEW

File: content/runner.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    // Listen for responses
+    this.port.onMessage.addListener((msg) => this.handleIncomingMessage(msg));
+  }
+
+  static execute (task, data) {
+    return Runner.get().execute(task, data);
+  }
+
+  execute (task, input) {
+    const id = uuid();
```
Comment (tomusdrw) [inline content/runner.js]: You should just use a number here, will be much faster.

File: content/runner.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    return new Promise((resolve, reject) => {
+      // Reject after no answer in 5s
```
Comment (tomusdrw) [inline content/runner.js]: Why is this needed? (also comment is not inline with timeout)

File: content/runner.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      // postMessage to the background script
+      this.port.postMessage({ id, data });
+    });
+  }
+
+  handleIncomingMessage (msg) {
+    let data;
+
+    try {
+      data = typeof msg === 'string'
```
Comment (tomusdrw) [inline content/runner.js]: Why is this needed? (`try`, `typeof`, etc)?

File: content/augmentor.js [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+    return element;
+  }
+
+  static augmentNode (key, node, resolved = {}) {
+    if (!node || node.getAttribute('data-parity-touched') === 'true') {
+      return;
+    }
+
+    node.setAttribute('data-parity-touched', true);
```
Comment (tomusdrw) [inline content/augmentor.js]: Would be good to extract this to constant.

File: content/augmentor.js [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+    title.className = styles.title;
+    title.innerHTML = nameHTML;
+    title.appendChild(addressElement);
+
+    const header = document.createElement('span');
+    header.className = styles.header;
+    header.appendChild(identityIcon);
+    header.appendChild(title);
+
+    element.innerHTML = `
```
Comment (tomusdrw) [inline content/augmentor.js]: I think we should avoid mixing `document.createElement` and string-concat DOM creation. Perhaps it would be best to use some template library?

File: content/augmentor.js [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+    const badge = Augmentor.getBadge(token, 32);
+    // Display with 3 decimals
+    const value = parseFloat(balance).toFixed(3);
+
+    return `
+      <span title="${name}" class="${styles.token}">
+        ${badge.outerHTML}
+        <span class="${styles.balance}">
+          <span class="${styles.value}">${value}</span>
+          <span class="${styles.tla}">${TLA}</span>
```
Comment (tomusdrw) [inline content/augmentor.js]: Possible injections here.

File: content/augmentor.js [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+
+  static getBadge (badge, height) {
+    const { src, title = '' } = badge;
+
+    const image = new Image();
+
+    image.src = src;
+    image.title = title.toUpperCase();
+    image.className = styles.badge;
+    image.style = `height: ${height}px;`;
```
Comment (tomusdrw) [inline content/augmentor.js]: Should use `style.height` instead.

File: content/augmentor.js [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    addressElement.className = styles.address;
+    addressElement.title = address;
+    addressElement.innerText = address;
+
+    let startedClick = false;
+    let doubleClicked = false;
+    let clickTimeout = null;
+
+    // Prevent closing on double-click on address (to select it's value)
+    addressElement.addEventListener('click', (event) => {
```
Comment (tomusdrw) [inline content/augmentor.js]: `dblclick` handling should be closed to a separate function

File: content/augmentor.js [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+        Augmentor.positionNode(badgesElement, iconsElement, { scale: 1.5, forceBottom: true });
+        Augmentor.positionNode(cardElement, iconsElement, { scale: 4 });
+      });
+  }
+
+  static positionNode (node, container, options = {}) {
+    const { scale = 1.5, offset = 5, forceBottom = false } = options;
+
+    const nodeRect = node.getBoundingClientRect();
```
Comment (tomusdrw) [inline content/augmentor.js]: This might break if augmented node is not visible, we should probably do positioning on `mouseenter`.

We may also save a lot of DOM nodes if we build those elements on `mouseenter` instead too.

File: content/augmentor.js [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+          }
+
+          const beforeText = node.innerText.slice(0, emailIndex);
+          const afterText = node.innerText.slice(emailIndex + email.length);
+
+          node.innerHTML = `${beforeText}<span>${email}</span>${afterText}`;
+          safeNode = node.querySelector('span');
+        }
+
+        if (!safeNode) {
```
Comment (tomusdrw) [inline content/augmentor.js]: It can never happen.

**Review by tomusdrw — CHANGES_REQUESTED** [review]
Grumbles :)

File: content/runner.js (hunk 2 above)
Comment (ngotchac) [reply content/runner.js]: It's just to make sure that the background script responds in a timely manner. If ever it is disconnected and whatnot, don't wait for the answer indefinitely.

File: content/runner.js (hunk 1 above)
Comment (ngotchac) [reply content/runner.js]: Was figuring it could be tricky if two content scripts from two different tabs would send a request with the same id. And it should really only send one request per page load (plus one after each mutations)

File: content/augmentor.js (hunk 8 above)
Comment (ngotchac) [reply content/augmentor.js]: It tried but couldn't get it to work... Will try again !

File: content/runner.js (hunk 3 above)
Comment (ngotchac) [reply content/runner.js]: Just to make sure it correctly parses the message, whether it's an object already or a JSON string

File: content/augmentor.js (hunk 10 above)
Comment (ngotchac) [reply content/augmentor.js]: True...

File: content/augmentor.js (hunk 9 above)
Comment (ngotchac) [reply content/augmentor.js]: Yeah, `mouseenter` and `mouseleave` wan't handling this effect as good as the CSS `hover` was.... Will see what would be the best.

File: content/runner.js (hunk 1 above)
Comment (tomusdrw) [reply content/runner.js]: Each `port` connection is separate in background thread, so there is no issue with collisions here.

Why do we send only one request? I thought we will send a request per match? If we wait for everything befoere doing anythnig it might take super long if there is a lot of addresses, I think it would be better to do it async for each element.

File: content/runner.js (hunk 2 above)
Comment (tomusdrw) [reply content/runner.js]: `postMessage` will throw if background thread is disconnected (https://developer.chrome.com/extensions/runtime#type-Port)

File: content/runner.js (hunk 3 above)
Comment (tomusdrw) [reply content/runner.js]: But why would it ever be a JSON string? IMHO would be better to always agree on the messages sent between content and background, less moving parts.

File: content/augmentor.js (hunk 9 above)
Comment (tomusdrw) [reply content/augmentor.js]: You can do `mouseenter` with initial animation (key-frame based) should have same effects as `:hover` (just make sure not to remove DOM nodes on `mouseleave` to early (or don't remove them at all))

Comment (ngotchac) [reply content/augmentor.js]: It was just _getting confused_ whenever entering/leaving/entering/etc. in a very short amount of time

File: content/runner.js (hunk 2 above)
Comment (ngotchac) [reply content/runner.js]: Ok

File: content/runner.js (hunk 3 above)
Comment (ngotchac) [reply content/runner.js]: Ok

File: content/runner.js (hunk 1 above)
Comment (ngotchac) [reply content/runner.js]: Ok. From what I tested there aren't that many different names/emails/whatever on a single page, and I was afraid the background would make twice or more the same call, but that's taken care of

File: content/augmentor.js (hunk 9 above)
Comment (tomusdrw) [reply content/augmentor.js]: I see, you may have some luck with `debounce` and using `display: inline-block` perhaps.

File: content/augmentor.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+      return node;
+    }
+
+    const emailIndex = node.innerText.indexOf(email);
+
+    if (emailIndex === -1) {
+      return;
+    }
+
+    const beforeText = node.innerText.slice(0, emailIndex);
```
Comment (tomusdrw) [inline content/augmentor.js]: Most probably should use `textContent` instead of `innerText` everywhere (see: https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent)

File: content/augmentor.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+    safeNode.insertAdjacentText('afterend', afterText);
+
+    // Replace the node with the safe node
+    node.parentElement.replaceChild(nextNode, node);
+
+    return safeNode;
+  }
+
+  static augmentNode (key, node, resolved = {}) {
+    if (!node || node.getAttribute(AUGMENTED_NODE_ATTRIBUTE) === 'true') {
```
Comment (tomusdrw) [inline content/augmentor.js]: We can probably just check `hasAttribute`

File: content/augmentor.js [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+      return;
+    }
+
+    const data = resolved[key];
+    node.setAttribute(AUGMENTED_NODE_ATTRIBUTE, true);
```
Comment (tomusdrw) [inline content/augmentor.js:90]: I wonder if we should set the attribute if data is not there, probably doesn't cause the issue cause we'll never call this function again (or maybe we will when `MutationObserver` launches and then new data can be fetched for that node?)

File: content/augmentor.js [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+            badges={ badges }
+            height={ height }
+            name={ name }
+            tokens={ tokens }
+          />
+        ));
+
+        const container = document.createElement('span');
+        container.className = styles.container;
+        container.appendChild(node.cloneNode(true));
```
Comment (tomusdrw) [inline content/augmentor.js]: I think it would be better to avoid cloning and `appendChild` after we replace it in `parentElement`, that way we don't remove any listeners that might be attached to this node.

File: content/augmentor.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+      const nextTop = pageHeight - offset - scaledClientRect.height;
+      node.style.top = `${nextTop}px`;
+    }
+
+    // If 5px of less of left border
+    if (scaledClientRect.top <= offset) {
+      if (forceBottom) {
+        const nextTop = scaledClientRect.height + 2 * offset + container.getBoundingClientRect().height;
+        node.style.top = `${nextTop}px`;
+        node.className += ` ${styles.fromBottom}`;
```
Comment (tomusdrw) [inline content/augmentor.js]: `classList.add`

Comment (tomusdrw) [reply content/augmentor.js]: (And probably should check if class is already there)

File: content/components/accountCard.js [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import { bind } from 'decko';
```
Comment (tomusdrw) [inline content/components/accountCard.js:17]: :+1: Neat :tada:!

File: content/components/augmentedIcon.js [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+export default class AugmentedIcon extends Component {
+
+  state = {
+    hover: false,
+    open: false
+  };
+
+  constructor (props) {
+    super(props);
+    this.setHover = debounce(this._setHover, 40, { leading: true });
```
Comment (tomusdrw) [inline content/components/augmentedIcon.js:51]: Have you seen that there is also `debounce` in `decko`?  Although you cannot configure it run on `leading` edge, right?

Comment (tomusdrw) [reply content/components/augmentedIcon.js:51]: I think it would be better to call it `setHoverDebounced` or something to distinguish it more from `this._setHover`)

File: content/extractor.js [hunk 18]
```diff
@@ -41,7 +42,7 @@ export default class Extractor {
     const matches = nodes.map((data) => {
       const { node, text } = data;
 
-      if (node.getAttribute('data-parity-touched') === 'true') {
+      if (node.getAttribute(AUGMENTED_NODE_ATTRIBUTE) === 'true') {
```
Comment (tomusdrw) [inline content/extractor.js]: `hasAttribute` suffices

**Review by tomusdrw — COMMENTED** [review]
Preact indeed looks awesome, can you include a screenshot/movie of what the experience looks like?
Left couple of minor grumbles as well :)

File: content/augmentor.js (hunk 11 above)
Comment (ngotchac) [reply content/augmentor.js]: Totally ! Didn't know this one.

File: content/augmentor.js (hunk 13 above)
Comment (ngotchac) [reply content/augmentor.js:90]: Well anyway the data is cached (had in mind to add a TTL) so calling it again within a few minutes won't change the result. I guess it just prevents trying multiple times for the same node

File: content/augmentor.js (hunk 15 above)
Comment (ngotchac) [reply content/augmentor.js]: That part I haven't refactored yet. Will come to this today !

File: content/augmentor.js (hunk 14 above)
Comment (ngotchac) [reply content/augmentor.js]: Figured out how to solve what I thought would be a style issue while answering you. So will fix !

File: content/components/augmentedIcon.js (hunk 17 above)
Comment (ngotchac) [reply content/components/augmentedIcon.js:51]: Yep agreed. And yes I know, but it's not as configurable as the `lodash` one, unfortunately...

File: content/augmentor.js (hunk 9 above)
Comment (ngotchac) [reply content/augmentor.js]: Solved it with `mousemove`. The `mouseleave` event was firing too often and randomly..

File: content/augmentor.js (hunk 13 above)
Comment (tomusdrw) [reply content/augmentor.js:90]: But "no data" is cached for that node right? And when mutation observer kicks in it may change node content to something that would match (so we should probably trigger fetching again), but it's an edge case, don't think we should catter for that.

Comment (ngotchac) [reply content/augmentor.js:90]: Ah yeah haven't thought of this one... Will keep it in mind !

Comment (ngotchac) [reply content/augmentor.js:90]: However it would be the same even if the value is matched: it won't reprocess the node whereas the inner data could have changed and thus the augmented icon wrong (or deleted)

**Comment by ngotchac** [comment]
Finally got it working correctly. Here's a demo : https://youtu.be/aJmo7O8gzOg

**Review by tomusdrw — APPROVED** [review]
